### PR TITLE
feat: Purchase Reconciliation Tool

### PR DESCRIPTION
**Pull request by vorasmit**
#75 continued here

### CODE REVIEW

File: india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+
+def _get_first_day(month, year):
+    """Returns first day of the month"""
+    return datetime.strftime(datetime(year, month, 1), "%Y-%m-%d")
+
+
+def _get_last_day(month, year):
+    """Returns last day of the month"""
+    return datetime(year, month, monthrange(year, month)[1])
+
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py]: These functions need to be consistent in what they return.

File: india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+    end_date = min(fiscal_year.end_date, _getdate(return_type))
+
+    # latest to oldest
+    return tuple(reversed(_get_periods(fiscal_year.start_date, end_date)))
+
+
+def _get_periods(start_date, end_date):
+    """Returns a list of month (formatted as `MMYYYY`) in given date range"""
+
+    return pd.date_range(start_date, end_date, freq="MS").strftime("%m%Y").tolist()
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py]: We need to remove `pandas` as a dependency (reasons: [1](https://github.com/frappe/erpnext/pull/30598#issuecomment-1090978603) [2](https://github.com/frappe/erpnext/issues/27047)). It is an overkill for this use-case.

Comment (sagarvora) [reply india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py]: I think this answer can achieve same result:
https://stackoverflow.com/a/28290050/4767738

```
import datetime
from dateutil.rrule import rrule, MONTHLY

strt_dt = datetime.date(2001,1,1)
end_dt = datetime.date(2005,6,1)

dates = [dt for dt in rrule(MONTHLY, dtstart=strt_dt, until=end_dt)]
```

File: india_compliance/gst_india/utils/gstr/gstr.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+def get_mapped_value(value, map):
+    return map.get(value)
+
+
+class GSTR:
+    # Maps of API keys to doctype fields
+    KEY_MAPS = frappe._dict()
+
+    # Maps of API values to doctype values
+    VALUE_MAPS = frappe._dict(
```
Comment (sagarvora) [inline india_compliance/gst_india/utils/gstr/gstr.py:18]: Shouldn't be a class var if it is designed to be mutated. Class vars are shared across all instances.

**Comment by vorasmit** [comment]
- [x] Search Index for link field in GST Inward Supply
- [x] Send as text to frontend (Frappe)
- [x] Don't pop list or lookup in the list
- [x] Flexibility for 2A download
- [x] Respect SEZ or Overseas in 2A or 2B Download
- [x] BOE in reco tool
- [ ] Create new invoice in new tab

**Comment by vorasmit** [comment]
continued in #916

**Pull request merged**